com.incquerylabs.v4md/src/main/com/incquerylabs/v4md/ViatraQueryAdapter.java

```diff
@@ -109,7 +109,9 @@ public AdvancedViatraQueryEngine getInitializedEngineChecked() {
 	 */
 	public void executeActionOnEngine(Consumer<AdvancedViatraQueryEngine> action) {
 		if(action != null) {
-			initializationActions.add(action);
+			synchronized (this) {
```

Comment:
I'd suggest an alternative approach here - instead of fine-grained synchronization, we should mark all engine initialization methods as synchronized (see https://docs.oracle.com/javase/tutorial/essential/concurrency/syncmeth.html). This is a more coarse-grained synchronization approach, but given that the underlying VIATRA query engine initialization is strictly single-threaded, this is safer, visible from the API definition and easier to maintain in the longer term.

To be clear, this also means we should remove all synchronized blocks if we synchronize the methods themselves.